### ISSUE TITLE
Non-backwards compatible change: CP using NLB

### ISSUE THREAD

**Issue by bellis-ai** [rancherfederal/rke2-aws-tf#90]
In a recent update, the control plane was changed to use a NLB instead of a classic load balancer. Those upgrading the module to the latest version will find the following error take place

- CP Loadbalancer is upgraded to NLB
- Target groups are created for the NLB
- Autoscaling group has "ignore_changes" on "load_balancer" and "target_groups" property, and ignores the changes for the latest load balancer.
- Because the autoscaling group does not have the correct load balancer settings, instances are not automatically assigned to the NLB and the control plane fails.

Not sure how to fix.

**Comment by bellis-ai** [comment]
Looks like it's a matter of just changing the Autoscaling group to use the new NLB, importing it back into state, and then adding the security groups (ends in -cp) to each control plane instance

**Comment by adamacosta** [comment]
I haven't yet worked out how to handle a migration gracefully to the NLB, but beware that if you just update in-place, the new load balancer will have a different DNS name from the old one and that will invalidate the server certificate being served by the Kube api-server, which will have the DNS of the old load balancer in its SAN list, placed in their automatically by our module.

I _believe_, but have not yet tested, that what you have to do is:

- Create the new NLB outside of Terraform first
- Grab the DNS name from AWS and add it to the TLS san list in the /etc/rke2/config.yaml files on all of the control plane nodes
- Cycle rke2 on the control plane nodes to generate a new certificate that will include this
- Import the load balancer into Terraform state
- Then do the rest of what you're saying above

Alternatively, if you have a custom URL and DNS record for the api-server and already included that in the TLS san list, none of this will matter.

**Comment by bellis-ai** [comment]
Thank you so much! I was just encountering this problem when trying to cycle out the old master nodes -- none were joining the cluster! I'll try this now.

**Comment by bellis-ai** [comment]
@adamacosta When you say
`Cycle rke2 on the control plane nodes to generate a new certificate that will include this`
What do you mean exactly? Restart the systemctl service? How do I cycle rke2? I am not very experienced in manual deployment of RKE2, so I'd like to know what I need to restart

**Comment by adamacosta** [comment]
Yes, run `systemctl restart rke2-server` on each control plane node. That _should_ generate a new certificate with the added TLS san for the new load balancer in it. Then new nodes should be able to join after that.

**Comment by bellis-ai** [comment]
So it looks like the changes are indeed propagated to serving-kube-apiserver.crt, but whatever cert is being used for the supervisor does not change out. I have no idea how to force change it.

**Comment by bellis-ai** [comment]
Figured it out. You have to invalidate the cached certificate data by deleting `/var/lib/rancher/rke2/server/tls/dynamic-cert.json`. No idea why this isn't done automatically when the certificate data is different.

**Comment by adamacosta** [comment]
Hey, thanks for figuring that out. Apologies for not following this better. I did get around to trying this out and it worked fine for me in terms of hitting the api server, but I only ran it on a single host, so the supervisor process would have been unused anyway. I'm not going to close this right away because we should put this in a real migration doc.
